numpy/_core/src/multiarray/stringdtype/casts.cpp

```diff
@@ -2072,70 +2243,83 @@ get_casts()
+
 #if NPY_SIZEOF_BYTE == NPY_SIZEOF_SHORT
-    casts[cast_i++] = StringToByteCastSpec;
-    casts[cast_i++] = ByteToStringCastSpec;
-    casts[cast_i++] = StringToUByteCastSpec;
-    casts[cast_i++] = UByteToStringCastSpec;
+    casts[cast_i++] = getStringToIntCastSpec<npy_byte,  npy_longlong,  NPY_BYTE>();
+    casts[cast_i++] = getStringToIntCastSpec<npy_ubyte, npy_ulonglong, NPY_UBYTE>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_byte,  long long,          NPY_BYTE>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_ubyte, unsigned long long, NPY_UBYTE>();
 #endif
 #if NPY_SIZEOF_SHORT == NPY_SIZEOF_INT
-    casts[cast_i++] = StringToShortCastSpec;
-    casts[cast_i++] = ShortToStringCastSpec;
-    casts[cast_i++] = StringToUShortCastSpec;
-    casts[cast_i++] = UShortToStringCastSpec;
+    casts[cast_i++] = getStringToIntCastSpec<npy_short,  npy_longlong,  NPY_SHORT>();
+    casts[cast_i++] = getStringToIntCastSpec<npy_ushort, npy_ulonglong, NPY_USHORT>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_short,  long long,          NPY_SHORT>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_ushort, unsigned long long, NPY_USHORT>();
 #endif
 #if NPY_SIZEOF_INT == NPY_SIZEOF_LONG
-    casts[cast_i++] = StringToIntCastSpec;
-    casts[cast_i++] = IntToStringCastSpec;
-    casts[cast_i++] = StringToUIntCastSpec;
-    casts[cast_i++] = UIntToStringCastSpec;
+    casts[cast_i++] = getStringToIntCastSpec<npy_int,  npy_longlong,  NPY_INT>();
+    casts[cast_i++] = getStringToIntCastSpec<npy_uint, npy_ulonglong, NPY_UINT>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_int,  long long,          NPY_INT>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_uint, unsigned long long, NPY_UINT>();
 #endif
 #if NPY_SIZEOF_LONGLONG == NPY_SIZEOF_LONG
-    casts[cast_i++] = StringToLongLongCastSpec;
-    casts[cast_i++] = LongLongToStringCastSpec;
-    casts[cast_i++] = StringToULongLongCastSpec;
-    casts[cast_i++] = ULongLongToStringCastSpec;
+    casts[cast_i++] = getStringToIntCastSpec<npy_longlong, npy_longlong,  NPY_LONGLONG>();
+    casts[cast_i++] = getStringToIntCastSpec<npy_longlong, npy_ulonglong, NPY_ULONGLONG>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_longlong, long long,          NPY_LONGLONG>();
+    casts[cast_i++] = getIntToStringCastSpec<npy_longlong, unsigned long long, NPY_ULONGLONG>();
 #endif
-    casts[cast_i++] = StringToDoubleCastSpec;
-    casts[cast_i++] = DoubleToStringCastSpec;
-    casts[cast_i++] = StringToFloatCastSpec;
-    casts[cast_i++] = FloatToStringCastSpec;
-    casts[cast_i++] = StringToHalfCastSpec;
-    casts[cast_i++] = HalfToStringCastSpec;
+
+    casts[cast_i++] = getStringToFloatCastSpec<npy_float16, NPY_HALF,  npy_half_isinf,      is_inf<double>, npy_double_to_half>();
+    casts[cast_i++] = getStringToFloatCastSpec<npy_float32, NPY_FLOAT, is_inf<npy_float32>, is_inf<double>, to_float<npy_float32>>();
```

Comment:
I don't think this `is_inf` function is necessary, just use `std::isinf` here.

Reply:
Originally I did this because `getStringToFloatCastSpec` took a `int (*)(NpyType)` since `npy_half_isinf` returns an `int` too.

I've changed this so that `getStringToFloatCastSpec` takes a `bool (*)(NpyType)` so that I can use `std::isinf` in these templates, but as a side effect I needed to write a _new_ wrapper function `as_bool` that casts the result of `npy_half_isinf` to `bool` for use in this template parameter. I would otherwise have used a lambda here but they're not supported in template parameters :(

Reply:
> I've changed this so that getStringToFloatCastSpec takes a bool (*)(NpyType) so that I can use std::isinf in these templates, but as a side effect I needed to write a new wrapper function as_bool that casts the result of npy_half_isinf to bool for use in this template parameter. I would otherwise have used a lambda here but they're not supported in template parameters :(

ugh... sorry, C++ is powerful but then stuff like this happens